### PR TITLE
Initial Sidebar UI implementation

### PR DESCRIPTION
**Pull request by simonhong**
Resolves https://github.com/brave/brave-browser/issues/12906

This PR only implements initial native sidebar UI.

Sidebar component manages persisted items list in pref store.
SidebarService does it and store items list is shared per profile.

In chrome layer, sidebar is implemented with MVC model.
They are SidebarController, SidebarModel and Sidebar and many views implementations.
Each browser window has its own MVC. So, BraveBrowser(subclass of Browser) owns them.
SidebarModel manages each item's runtime state and it observes SidebarService to get
notification about item add or delete.

Sidebar feature is disabled by default and can enable via brave://flags only on nightly/local build.
During the runtime, sidebar can be toggled via settings.

Also,`on mouseover` and `on click` in shows option are not implemented yet.

Note: Sidebar webviews don't work properly yet because their WebContentsDelegate is not set properly.
Implementing WebContentsDelegate will be done as a separate task.

**Known issues**
- Crashed when history item is clicked in private/tor window. I think loading history with incognito profile is not allowed.
- Need to think about how to handle bookmark/history item in private/tor window. Loading them in private/tor window tab are redirected to normal window.

---------------
Settings option to show/hide during the runtime
image

Feature switch is only visible on nightly and disabled by default
image

Sidebar items overflow mode - sidebar items can be scrolled via up/down arrow button or mouse scrolling.
image

Sidebar Item add bubble menu (bubble should have shadowed border)
image

Item added feedback bubble (bubble should have shadowed border)
image

Item focused
image

Dark mode/Private/Private with Tor window
image
image
image

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
`npm run test brave_unit_tests -- --filter=*Sidebar*`
`npm run test brave_browser_tests -- --filter=*Sidebar*`

### CODE REVIEW

**Comment by bsclifton** [comment]
Just starting to check out - will share notes / findings here 😄 

1. Noticed that if the info bar shows, it will cover up the sidebar. Not sure if we wanted to handle this case (and move it down) or leave it as-is:
image

**Comment by simonhong** [comment]
@bsclifton no.4 is my intended behavior to not create multiple wallet tabs in the same window.
It just activates if current window already has opened wallet tab.

**Comment by bsclifton** [comment]
@simonhong RE: number 4 above, definitely makes sense 😄 We can solve that in the future

Some more notes:

1. Just noticed this (on right click)- nice touch 😄 Doesn't seem to do anything yet:
image

2. When resizing the window (with sidebar enabled or disabled), I'm seeing the window be filled with blue before it gets painted. I need to try master to verify if this is a problem with master or if it was introduced with any theme changes here (just noticed now). For reference, I was resizing window vertically very fast (on Windows)

Very nice work 😄 Will have to check out code next

**Comment by simonhong** [comment]
@bsclifton `Show Sidebar` is not yet implemented. I'm working on `item overflow mode` and `show sidebar option` now. 
Both feature will get merged soon to this PR 😄

**Comment by simonhong** [comment]
We need another bubble border that draws arrow. Working on it.
Upstream `BubbleBorder` doesn't draw arrow. Although It uses `arrow` argument, it just use arrow args for bubble positioning.

**Review by karenkliu — CHANGES_REQUESTED** [review]
@simonhong thanks for sending this over!!! it looks great. One tiny tweak - could we give the line items in the "Add sidebar" menu a bit more space? The line height looks a little tight. 

image

**Review by karenkliu — APPROVED** [review]

**Comment by simonhong** [comment]
Kindly ping :)

**Comment by diracdeltas** [comment]
some comments/concerns about loading remote side panel content (like https://together.brave.com) in a webview:

* the webview doesn't have navigation controls so you can get stuck on a page. for instance after clicking the link to the privacy policy in together.brave.com:
Screen Shot 2021-01-28 at 2 53 50 PM
* i think some security conscious users will complain that they can't see the URL of the webview or inspect it using devtools AFAICT
* are webviews subject to the same security/privacy restrictions as if the same URL were loaded in a tab? for instance are shields applied to them? what about browser extensions?
* if i open the side panel in a private tab, does it load using the associated non-OTR profile? (so the cookies from the regular profile will get sent in requests in the webview)
* if i open the brave together side panel in a Tor tab, i assume it won't be loaded using Tor? this might be confusing to users.

**Comment by simonhong** [comment]
@diracdeltas Thanks for checking!
I think I should do more for webcontents in the sidebar webview.
Maybe implementing WebContentsDelegate, Attaching vairous TabHelpers and etc.
Need to check all the things that tab's webview has.

Sidebar webview uses OTR-profile if it's in private(tor) window.
I also think some panel needs navigation UI(cc @karenkliu)

This PR only focus on implementing initial native UI.
Although this is only enabled for nightly and disabled by default, how about disabling builtin items(that creates webview) in this PR and enable it with f/u(with handling of above items that @diracdeltas mentioned above)?
WDYT? @rebron @karenkliu

**Review by bsclifton — APPROVED** [review]
Tested this out fairly well- also got a chance to go through the code. Things look good 😄👍 

Very nice work!

**Comment by iefremov** [comment]
@simonhong Sorry for the dealy, I'll go through the PR on Monday

**Comment by simonhong** [comment]
@iefremov NP. thanks!

**Comment by karenkliu** [comment]
@simonhong 
>Although this feature is only added for nightly and disabled by default, how about disabling builtin items(that creates webview) in this PR and enable it with f/u(with handling of above items that @diracdeltas mentioned above)?

SGTM. Also yeah, we know there are some navigation and interaction issues that still need to be figured out. Waiting for the prototype to land in Nightly so I can test it, figure out the issues, and come up with some designs for it. (At this point, it's hard to figure out what needs to be done by looking at static designs).

Thanks for all your hard work!!

**Comment by diracdeltas** [comment]
@karenkliu you can download a macos build from https://brave-jenkins-build-artifacts.s3.amazonaws.com/brave-browser-build-pr/sidebar/6e9e993924b30330c744cad61b5ac619edc57471-8a0c7739cf4568183c54d0659d1fc115e89f6f96/macos/brave-v1.21.24-darwin-x64.zip (may require VPN)

**Comment by petemill** [comment]
I think it's fine to merge with those issues since this is behind a feature flag, and therefore is designed to gather exactly these kinds of usability issues from a slightly wider audience.

Having said that, do we really want or need sidebar at all on any Private windows? We could just disable in those windows, at least for now.

**Comment by karenkliu** [comment]
>Having said that, do we really want or need sidebar at all on any Private windows? We could just disable in those windows, at least for now.

@simonhong I suggested that the sidebar be in its out-of-the-box state for every new private window/with tor, but this is also a good idea;   let's do whatever is easier to implement.

**Comment by simonhong** [comment]
@karenkliu  Temporarily disabled sidebar for private profile(tor/guest).
Sidebar in private window has crash issue. (As I mentioned above, can't load history with private profile)
@diracdeltas I didn't disable sidebar panel because user can't add new items that target for sidebar webview.
Currently only our default items(brave together, history and bookmark) can be loaded.
I think having sidebar panel(webview) opening feature is needed to develp UI more.

**Comment by simonhong** [comment]
Kindly ping to @brave/patch-reviewers :)

File: patches/chrome-browser-ui-browser.cc.patch [hunk 1]
```diff
@@ -1,7 +1,16 @@
...
-index 2c9d816e1aff340615061e888cbbcd69300befcd..76f9bc4537210c08f0d696f6f41ec7b9a08a943e 100644
+index 2c9d816e1aff340615061e888cbbcd69300befcd..ff9c60969b10b9037097ea0186cc35c96f0a5a33 100644
 --- a/chrome/browser/ui/browser.cc
 +++ b/chrome/browser/ui/browser.cc
+@@ -434,7 +434,7 @@ Browser::BrowserCreationStatus Browser::GetBrowserCreationStatusForProfile(
+ Browser* Browser::Create(const CreateParams& params) {
+   CHECK_EQ(BrowserCreationStatus::kOk,
+            GetBrowserCreationStatusForProfile(params.profile));
+-  return new Browser(params);
++  return new BraveBrowser(params);
```
Comment (mkarolin) [inline patches/chrome-browser-ui-browser.cc.patch:10]: Maybe #define this as BRAVE_BROWSER_CREATE in the `chromium_src` override and insert it before the chromium's `return` statement?

Comment (simonhong) [reply patches/chrome-browser-ui-browser.cc.patch]: Fixed.

File: chromium_src/chrome/browser/unexpire_flags.cc [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#include "../../../../chrome/browser/unexpire_flags.cc"
+#undef IsFlagExpired
+
+namespace flags {
+
+bool IsFlagExpired(const flags_ui::FlagsStorage* storage,
+                   const char* internal_name) {
+  // Enable sidebar feature only for nightly/development.
+  version_info::Channel channel = chrome::GetChannel();
+  if (strncmp("sidebar", internal_name, 7) == 0 &&
```
[the remark this answers is not in the document]
Comment (simonhong) [reply chromium_src/chrome/browser/unexpire_flags.cc]: Right, fixed to compare more precisely with `LowerCaseEqualsASCII`.
Thanks for catch!

**Review by mkarolin — APPROVED** [review]
`patches` and `chromium_src` LGTM

**Comment by simonhong** [comment]
Merged because only known test (also reported) are failed.

**Comment by BroJenuel** [comment]
This sidebar would be awesome XD

**Comment by CrazeXD** [comment]
Are we planning to get this feature any time soon? Really would appreciate.

**Pull request merged**